### PR TITLE
Optimize radix tree code

### PR DESCRIPTION
**Pull request by platfowner**
Insights from experiments:
- Cloning cost is critical -> Introduced radix node versioning and cloning (with parent state node)
- Child node counting is critical -> Introduced numTerminalNodes
- Parent node counting is critical -> Adopt optimized parent counting functions
- Dynamic radix tree is not very helpful -> Deprecated for code simplicity and consistency
- Terminal radix node cache was not helpful
- Array radix child map was not helpful
- Radix tree with hashed label (with simple hash function) was not very helpful
- Epoch calibration was not very helpful
- Radix tree with longer label radix (e.g. length = 2) was not helpful
- Radix tree with other radix bases (e.g. base-4) was not helpful

Change summary:
- Add radix node versioning and cloning by multiple parents with enableRadixNodeVersioning flag
- Fix bugs in RadixNode._mergeToChild() and setStateNode()
- Fix child cloning bug in getRadixNodeForSetting()
- Optimize parent counting using StateNode.hasAParent() and StateNode.hasMultipleParents()
- Simplify code by deprecating dynamic radix tree
- Simplify code by deprecating enableStateVersionOpt flag

Related issues:
https://github.com/ainblockchain/ain-blockchain/issues/519
https://github.com/ainblockchain/ain-blockchain/issues/517

Performance test result: ~25 tps
(http://35.189.163.153:3000/d/M6YpMSyMz/blockchain-servers?from=1631874163694&to=1631875324341&orgId=1)

### CODE REVIEW

File: db/radix-tree.js [hunk 1]
```diff
@@ -113,143 +138,224 @@ class RadixTree {
...
           `at: ${new Error().stack}.`);
       // Does nothing.
       return node;
     }
-    const parent = node.getParent();
+    const theOnlyChild = node.getChildNodes()[0];
+    if (theOnlyChild.numParents() > 1) {
+      // Cannot merged!!
+      // Does nothing.
+      return [node];
```
Comment (liayoo) [inline db/radix-tree.js:256]: should other return values also be arrays or should this be just node?

File: db/radix-tree.js [hunk 2]
```diff
@@ -113,143 +138,224 @@ class RadixTree {
...
-      parent.deleteChild(node.getLabelRadix());
-      if (parent.numChildren() === 1 && !parent.hasStateNode() && parent.hasParent()) {
-        nodeToUpdateProofHash = this._mergeToChild(parent);
+      } else {  // the node has only 1 parent.
+        const theOnlyParent = node.getParentNodes()[0];
+        theOnlyParent.deleteChild(labelRadix);  // delete child!
+        nodesToUpdate.push(theOnlyParent);
+        if (theOnlyParent.numChildren() === 1 &&  // the parent has only 1 child after deletion.
+            !theOnlyParent.hasChildStateNode() &&  // the parent has no state node
+            theOnlyParent.numParents() !== 0) {  // the parent is not a root.
```
Comment (liayoo) [inline db/radix-tree.js:314]: question) Could the last condition be changed to theOnlyParent.numParents() === 1?

File: db/radix-tree.js [hunk 3]
```diff
@@ -283,56 +389,61 @@ class RadixTree {
...
+  }
+
   /**
    * Converts the tree to a javascript object.
    * This is for testing / debugging purpose.
    */
-  toJsObject(withProofHash = false, withStateNodeDetails = false) {
-    return this.root.toJsObject(withProofHash, withStateNodeDetails);
+  toJsObject(
+      withVersion = false, withProofHash = false, withTreeInfo = false, withNumParents = false) {
```
Comment (liayoo) [inline db/radix-tree.js:445]: ditto (options as an object)

File: db/radix-node.js [hunk 4]
```diff
@@ -386,37 +529,66 @@ class RadixNode {
...
+      const childStateNode = this.getChildStateNode();
+      childStateNode.deleteParent(parentStateNodeToDelete);
     }
     this.reset();
     numAffectedNodes++;
 
     return numAffectedNodes;
   }
 
+  deleteRadixTreeVersion() {
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/radix-node.js:503]: As in the case of state tree, radix tree version here means the radix nodes belong to only a specific root, and radix tree version deletion means deleting all of them. This is check with 'node.numParents() > 0' logic.

File: db/radix-node.js [hunk 5]
```diff
@@ -386,37 +529,66 @@ class RadixNode {
...
+
+    return numAffectedNodes;
+  }
+
   /**
    * Converts the subtree to a js object.
    * This is for testing / debugging purpose.
    */
-  toJsObject(withProofHash = false, withTreeInfo = false) {
+  toJsObject(withVersion = false, withProofHash = false, withTreeInfo = false, withNumParents = false) {
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/radix-node.js:525]: This is testing purpose only, and we can keep the simplicity of expressions with the current form?
```
      assert.deepEqual(nodeAnother.toJsObject(true, true, false, true), {
        ".num_parents": 0,
        ".radix_ph": null,
        ".radix_version": null,
      });
```
vs.
```
      assert.deepEqual(nodeAnother.toJsObject({
        withVersion: true,
        withProofHash: true,
        withTreeInfo: false,
        withNumParents: true
      }), {
        ".num_parents": 0,
        ".radix_ph": null,
        ".radix_version": null,
      });
```

File: db/radix-tree.js [hunk 6]
```diff
@@ -11,9 +11,17 @@ const RadixNode = require('./radix-node');
...
+  constructor(version = null, parentStateNode = null) {
+    this.root = new RadixNode(version, parentStateNode);
+    this.numTerminalNodes = 0;
+  }
+
+  clone(version, parentStateNode) {
+    const clonedTree = new RadixTree();
+    clonedTree.root = this.root.clone(version, parentStateNode);
+    clonedTree.numTerminalNodes = this.numTerminalNodes;
+    // NOTE(platfowner): terminalNodeCache is not copied.
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/radix-tree.js]: Just removed. Thanks!

File: db/radix-tree.js [hunk 7]
```diff
@@ -45,66 +53,83 @@ class RadixTree {
...
-    if (terminalNode === undefined) {
-      return null;
-    }
-    return terminalNode;
-  }
+    const radixLabel = RadixTree._toRadixLabel(stateLabel);
+    let curNode = this.root
+    let labelIndex = 0;
+    while (labelIndex < radixLabel.length) {
+      const labelRadix = radixLabel.charAt(labelIndex);
```
[the remark this answers is not in the document]
Comment (platfowner) [reply db/radix-tree.js]: Just renamed: labelRadix -> childLabelRadix as its value is used as the label radix of the child node. Thanks!

File: db/radix-tree.js (hunk 1 above)
Comment (platfowner) [reply db/radix-tree.js:255]: Just fixed. Thanks!

File: db/radix-tree.js (hunk 2 above)
Comment (platfowner) [reply db/radix-tree.js]: Right, they're logically equivalent. Just changed so. Thanks!

File: db/radix-tree.js (hunk 3 above)
Comment (platfowner) [reply db/radix-tree.js:432]: ditto

**Comment by platfowner** [comment]
Comments addressed. PTAL~ @liayoo 
Additionally, just simplified the code by adopting enableRadixNodeVersioning=true logic all the time.

**Comment by liayoo** [comment]
question) This issue started occurring before this PR, but do you think it's possible that when a node is restarted, the radix tree node orders can be changed, which may result in state proof hash mismatch? I've been seeing the mismatch issue (along with some consensus-related errors) from time to time with the HE sharding & sharding integration tests and I've been meaning to debug it.. I'm suspecting the error happens after the node restarting test cases. Curious to hear what you think about it!

**Comment by platfowner** [comment]
> question) This issue started occurring before this PR, but do you think it's possible that when a node is restarted, the radix tree node orders can be changed, which may result in state proof hash mismatch? I've been seeing the mismatch issue (along with some consensus-related errors) from time to time with the HE sharding & sharding integration tests and I've been meaning to debug it.. I'm suspecting the error happens after the node restarting test cases. Curious to hear what you think about it!

Ah, this reminds me of some TODOs in radix-tree.js, which I was thinking to address right after this PR. I'll work on it soon. thanks!
fyi, we need our own feature for that as we cannot piggyback on hash map any more.

File: db/radix-node.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply db/radix-node.js:503]: Ah, I see. Thanks for the explanation!

File: db/radix-node.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply db/radix-node.js:525]: Right, in that case, it looks good!

**Comment by liayoo** [comment]
> > question) This issue started occurring before this PR, but do you think it's possible that when a node is restarted, the radix tree node orders can be changed, which may result in state proof hash mismatch? I've been seeing the mismatch issue (along with some consensus-related errors) from time to time with the HE sharding & sharding integration tests and I've been meaning to debug it.. I'm suspecting the error happens after the node restarting test cases. Curious to hear what you think about it!
> 
> Ah, this reminds me of some TODOs in radix-tree.js, which I was thinking to address right after this PR. I'll work on it soon. thanks!
> fyi, we need our own feature for that as we cannot piggyback on hash map any more.

Right, right! Let's include these in our q4 items then :)

File: db/state-node.js [hunk 8]
```diff
@@ -248,24 +263,41 @@ class StateNode {
...
+    if (FeatureFlags.enableRadixTreeLayers) {
+      return RadixTree.getParentStateNodes(this.getParentRadixNodes())
+    } else {
+      return Array.from(this.parentSet);
+    }
   }
 
-  numParents() {
-    return this.parentSet.size;
+  hasAParent() {
```
Comment (liayoo) [inline db/state-node.js:273]: minor suggestion) hasSingleParent ?

**Review by liayoo — APPROVED** [review]
LGTM with a minor comment. Thanks!!

File: db/state-node.js (hunk 8 above)
Comment (platfowner) [reply db/state-node.js]: Just renamed. Thanks!

**Comment by platfowner** [comment]
Thanks for the review, now merging..

**Pull request merged**